src/Toolchain.ml

```diff
@@ -0,0 +1,386 @@
+    | x :: xs -> (
+      match x with
+      | Global -> None
+      | Esy root
+      | Opam root -> (
+        match ofName ~root name with
+        | Some y ->
+          if compare x y = 0 then
+            Some x
+          else
+            findByName name xs
+        | None -> None ) )
+
+  let makeEsy root = Esy root
+
+  let makeOpam root = Opam root
+end
+
+module PackageManagerSet : sig
+  include Set.S with type elt = PackageManager.t
+end =
+  Set.Make (PackageManager)
+
+let packageManagerSetOfResultList ~debugMsg lst =
+  Js.Console.info debugMsg;
+  match lst with
+  | Ok lst ->
+    List.iter (fun x -> x |> PackageManager.toString |> Js.Console.info) lst;
+    lst |> PackageManagerSet.of_list
+  | Error msg ->
+    Js.Console.error2
+      (Printf.sprintf "Error during extracting %s", debugMsg)
+      msg;
+    PackageManagerSet.empty
+
+type resources =
+  { cmd : Cmd.t
+  ; kind : PackageManager.t
+  ; projectRoot : Fpath.t
+  }
+
+let makeResources ~env ~kind ~projectRoot =
+  Cmd.make ~env ~cmd:(PackageManager.toCmdString kind)
+  |> P.then_ (function
+       | Error msg -> Error msg |> P.resolve
+       | Ok cmd -> Ok { cmd; kind; projectRoot } |> P.resolve)
+
+let ofPackageManagerName ~env ~name ~projectRoot ~toolchainRoot =
+  match PackageManager.ofName ~root:toolchainRoot name with
+  | Some kind -> makeResources ~env ~kind ~projectRoot
+  | None -> Error ("Invalid package manager name: " ^ name) |> P.resolve
+
+let parseOpamEnvOutput opamEnvOutput =
+  opamEnvOutput |> Js.String.split "\n"
+  |. Belt.Array.keepMap (fun r ->
+         match r |> Js.String.split "=" |> Array.to_list with
+         | [ k; v ] -> Some (k, v)
+         | [] ->
+           (* TODO Environment entries are not necessarily key value pairs *)
+           Js.Console.info "Splitting on '=' in env output failed";
+           None
+         | l ->
+           Js.Console.info2
+             "Splitting on '=' in env output returned more than two items: "
+             (l |> Array.of_list |> Js.Array.joinWith ",");
+           None)
+  |> Js.Dict.fromArray |> Result.return |> P.resolve
+
+let env ~cmd ~kind =
+  match kind with
+  | PackageManager.Global -> Ok Process.env |> P.resolve
+  | Esy root ->
+    Cmd.output cmd
+      ~args:[| "command-env"; "--json"; "-P"; Fpath.toString root |]
+      ~cwd:(Fpath.toString root)
+    |> P.mapOk (fun stdout ->
+           match Json.parse stdout with
+           | Some json ->
+             json
+             |> Json.Decode.dict Json.Decode.string
+             |> Result.return |> P.resolve
+           | None ->
+             Error ("'esy command-env' returned non-json output: " ^ stdout)
+             |> P.resolve)
+  | Opam root ->
+    Cmd.output cmd ~args:[| "exec"; "env" |] ~cwd:(Fpath.toString root)
+    |> P.mapOk parseOpamEnvOutput
+
+let supportedPackageManagers ~env ~root =
+  let supportedPackageManagers =
+    [ PackageManager.Esy root; Esy Fpath.(root / ".vscode" / "esy"); Opam root ]
```

Comment:
This path calculated three times. Maybe we can put it to `Utils` or some other shared place?

Reply:
Yes, indeed. Thanks for addressing this in the commits too